azafea/event_processors/endless/metrics/v3/model/__init__.py

```diff
@@ -0,0 +1,625 @@
+    kernel = Column(BigInteger, nullable=False)
+    #: time spent in the initrd (if known) in µsec
+    initrd = Column(BigInteger, nullable=False)
+    #: time spent in userspace in µsec
+    userspace = Column(BigInteger, nullable=False)
+    #: total time spent to boot in µsec
+    total = Column(BigInteger, nullable=False)
+
+    @staticmethod
+    def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
+        return {
+            'firmware': clamp_to_int64(payload.get_child_value(0).get_uint64()),
+            'loader': clamp_to_int64(payload.get_child_value(1).get_uint64()),
+            'kernel': clamp_to_int64(payload.get_child_value(2).get_uint64()),
+            'initrd': clamp_to_int64(payload.get_child_value(3).get_uint64()),
+            'userspace': clamp_to_int64(payload.get_child_value(4).get_uint64()),
+            'total': clamp_to_int64(payload.get_child_value(5).get_uint64()),
+        }
+
+
+class ComputerInformation(SingularEvent):
+    """Information about the computer RAM, disk and CPU.
+
+    Sent at most once a day.
+
+    :UUID name: TODO
+
+    .. versionadded:: 4.0.0
+
+    """
+    __tablename__ = 'computer_information_v3'
+    __event_uuid__ = '81f303aa-448d-443d-97f9-8d8a9169321c'
+    __payload_type__ = '(uuuua(sqd))'
+
+    #: total RAM size in mebibytes (2^20)
+    total_ram = Column(BigInteger, nullable=False)
+    #: total disk space in gibibytes (2^30)
+    total_disk = Column(BigInteger, nullable=False)
+    #: used disk space in gibibytes (2^30)
+    used_disk = Column(BigInteger, nullable=False)
+    #: free disk space in gibibytes (2^30)
+    free_disk = Column(BigInteger, nullable=False)
+    #: array of CPU model (e.g. Intel(R) Core(TM) i7-5500U CPU @ 2.40GHz),
+    #: number of cores (e.g. 4) and maximum CPU speed in MHz or current CPU speed
+    #: if maximum can’t be determined (e.g. 3000.0)
+    info = Column(JSONB, nullable=False)
+
+    @staticmethod
+    def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
+        cpu_info = []
+        for i in range(payload.get_child_value(4).n_children()):
+            item = payload.get_child_value(i)
+            cpu_info.append({
+                'model': item.get_child_value(0).get_string(),
+                'cores': item.get_child_value(1).get_uint16(),
+                'max_frequency': item.get_child_value(2).get_double(),
+            })
+        return {
+            'total_ram': payload.get_child_value(0).get_uint32(),
+            'total_disk': payload.get_child_value(1).get_uint32(),
+            'used_disk': payload.get_child_value(2).get_uint32(),
+            'free_disk': payload.get_child_value(3).get_uint32(),
+            'cpu_info': cpu_info,
+        }
+
+
+# -- Aggregate events ---------------------------------------------------------
+
+class TimeSpentByForegroundApp(AggregateEvent):
+    """Number of seconds spent by an application open in foreground.
+
+    Aggregation is done by day.
```

Comment:
We have to be sure about the aggregation periods.

Reply:
We discussed yesterday using two fixed aggregation periods, calendar daily and calendar monthly, for these three aggregate events, and hence having 2 × 3 = 6 tables. Each has a unique UUID and the period is a property of that UUID. Is there a good PostgreSQL column type for "calendar day" / "calendar month"?